File: java/ValidParenthesis.java
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

// https://neetcode.io/problems/validate-parentheses?list=neetcode150
class ValidParenthesis {
    public boolean isValid(String str) {
        if (str.length() <= 1) {
            return false;
        }

        Stack<Character> stack = new Stack<>();
        var brackets = new HashMap<Character, Character>();
        brackets.put('[', ']');
        brackets.put('{', '}');
        brackets.put('(', ')');


        for (int i = 0; i < str.length(); i++) {
            var bracket = str.charAt(i);
            if (isOpeningBracket(bracket)) {
                stack.push(bracket);
            } else {
                // this means we only have closing brackets
                // in stack
                if (stack.size() == 0) {
                    return false;
                }
                // the current bracket we are iterting over
                // must match the stack head
                var lastBracket = stack.pop();
                var closingBracket = brackets.get(lastBracket);
                if (bracket != closingBracket) {
                    return false;
                }
            }
        }

        return stack.size() == 0;
    }

    private boolean isOpeningBracket(char c) {
        return c == '[' || c == '{' || c == '(';
    }
}

File: go/valid_parenthesis.go
// https://leetcode.com/problems/valid-parentheses/

func isValid(s string) bool {
    if len(s) <= 1 {
        return false
    }
    
    closeToOpen := make(map[rune]rune)
    closeToOpen[']'] = '['
    closeToOpen['}'] = '{'
    closeToOpen[')'] = '('

    stack := []rune{}

    for _, char := range s {
        if open, exists := closeToOpen[char]; exists {
            // If it's a closing bracket, check stack
            // Open bracket must match the previous element in slice
            if len(stack) == 0 || open != rune(stack[len(stack)-1]) {
                return false
            }
            
            // pop from stack
            stack = stack[:len(stack)-1]
        } else {
            // It's an opening bracket, push to stack
            stack = append(stack, char)
        }
    }

    return len(stack) == 0
}
